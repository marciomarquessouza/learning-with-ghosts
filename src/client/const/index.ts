export * from './actions'
export * from './meshes'
export * from './armatures'
export * from './params'
