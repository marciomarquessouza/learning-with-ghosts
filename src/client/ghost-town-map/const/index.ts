export * from './actions'
export * from './meshes'
export * from './armatures'
export * from './params'
export * from './colors'
export * from './elements'
export * from './characters'
export * from './languages'
export * from './interactions'
