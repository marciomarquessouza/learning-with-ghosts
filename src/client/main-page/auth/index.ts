export { default as logInWithEmailAndPassword } from './logInWithEmailAndPassword'
export { default as logout } from './logout'
export { default as registerWithEmailAndPassword } from './registerWithEmailAndPassword'
export { default as sendPasswordReset } from './sendPasswordReset'
export { default as signInWithGoogle } from './signInWithGoogle'
export { default as signInWithFacebook } from './signInWithFacebook'
export { default as SignInWithTwitter } from './SignInWithTwitter'
