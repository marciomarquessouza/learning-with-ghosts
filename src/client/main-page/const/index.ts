export const GITHUB_URL = 'https://github.com/marciomarquessouza/learning-with-ghosts'
