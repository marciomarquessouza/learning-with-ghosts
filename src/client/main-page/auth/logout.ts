import { signOut } from 'firebase/auth'
import { auth } from '../config/firebase'

function logout() {
    signOut(auth)
}

export default logout
