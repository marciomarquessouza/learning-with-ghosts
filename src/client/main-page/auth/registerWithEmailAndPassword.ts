import { createUserWithEmailAndPassword } from 'firebase/auth'
import { addDoc, collection } from 'firebase/firestore'
import { auth, db } from '../config/firebase'

async function registerWithEmailAndPassword(name: string, email: string, password: string) {
    try {
        const res = await createUserWithEmailAndPassword(auth, email, password)
        const user = res.user
        await addDoc(collection(db, 'users'), {
            uid: user.uid,
            name,
            authProvider: 'local',
            email,
        })
    } catch (err) {
        throw new Error('Error to register with email and password')
    }
}

export default registerWithEmailAndPassword
