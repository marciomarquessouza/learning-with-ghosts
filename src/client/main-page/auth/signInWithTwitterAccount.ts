import { TwitterAuthProvider, signInWithPopup } from 'firebase/auth'
import { query, getDocs, collection, where, addDoc } from 'firebase/firestore'
import { auth, db } from '../config/firebase'

const twitterProvider = new TwitterAuthProvider()

export async function signInWithTwitter() {
    try {
        const response = await signInWithPopup(auth, twitterProvider)
        const user = response.user
        const q = query(collection(db, 'users'), where('uid', '==', user.uid))
        const docs = await getDocs(q)
        if (docs.docs.length === 0) {
            await addDoc(collection(db, 'users'), {
                uid: user.uid,
                name: user.displayName,
                authProvider: 'twitter',
                email: user.email,
            })
        }
    } catch (err) {
        throw new Error('Error to sign in with a Twitter Account')
    }
}
