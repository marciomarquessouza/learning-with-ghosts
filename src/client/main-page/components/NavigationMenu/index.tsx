import React from 'react'
import { Link } from 'react-router-dom'
import { User } from 'firebase/auth'
import DropdownMenuAvatar from '../DropdownMenuAvatar'

export interface NavigationMenuProps {
    user?: User | null
    onLogout: () => void
}

function NavigationMenu({ user, onLogout }: NavigationMenuProps) {
    return (
        <div className="flex flex-row items-center">
            <div className="visible md:invisible">
                <svg
                    height="32px"
                    id="Layer_1"
                    version="1.1"
                    viewBox="0 0 32 32"
                    className="w-32"
                    fill="white"
                >
                    <path d="M4,10h24c1.104,0,2-0.896,2-2s-0.896-2-2-2H4C2.896,6,2,6.896,2,8S2.896,10,4,10z M28,14H4c-1.104,0-2,0.896-2,2  s0.896,2,2,2h24c1.104,0,2-0.896,2-2S29.104,14,28,14z M28,22H4c-1.104,0-2,0.896-2,2s0.896,2,2,2h24c1.104,0,2-0.896,2-2  S29.104,22,28,22z" />
                </svg>
            </div>
            <div className="invisible md:visible flex flex-row justify-items-center items-center">
                <div className="mx-4">
                    <a
                        href="https://github.com/marciomarquessouza/learning-with-ghosts"
                        target="_blank"
                    >
                        <span className="font-josefin font-bold text-lg text-white">ABOUT</span>
                    </a>
                </div>
                <div className="mx-4">
                    {user ? (
                        <DropdownMenuAvatar
                            name={user.displayName || 'User'}
                            email={user.email}
                            photoUrl={user.photoURL}
                            onLogout={onLogout}
                        />
                    ) : (
                        <Link to="/login">
                            <button className="bg-primary hover:bg-primary-dark py-2 px-4 rounded">
                                <span className="font-josefin font-bold text-lg text-white">
                                    LOGIN
                                </span>
                            </button>
                        </Link>
                    )}
                </div>
            </div>
        </div>
    )
}

export default NavigationMenu
