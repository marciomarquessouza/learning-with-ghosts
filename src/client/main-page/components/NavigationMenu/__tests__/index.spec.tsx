import React from 'react'
import { User } from 'firebase/auth'
import { render, fireEvent } from '@testing-library/react'

import NavigationMenu from '..'
import { GITHUB_URL } from '../../../const'

jest.mock('react-router-dom', () => ({
    ...(jest.requireActual('react-router-dom') as any),
    Link: () => <a />,
}))

describe('components >> NavigationMenu', () => {
    const defaultMockedProps = {
        user: undefined,
        onLogout: jest.fn(),
        onLogin: jest.fn(),
    }

    it('should render the NavigationMenu component properly', () => {
        const { container } = render(<NavigationMenu {...defaultMockedProps} />)

        expect(container).toMatchSnapshot('NavigationMenu-default')
    })

    it('should render the NavigationMenu component properly when the user is signed', () => {
        const mockedProps = { ...defaultMockedProps, user: { uid: '__MOCK__ID__' } as User }
        const { container } = render(<NavigationMenu {...mockedProps} />)

        expect(container).toMatchSnapshot('NavigationMenu-user')
    })

    it('should should fire the window event open when the ABOUT link is pressed', () => {
        global.open = jest.fn()
        const mockedProps = { ...defaultMockedProps }
        const { getByText } = render(<NavigationMenu {...mockedProps} />)

        fireEvent.click(getByText('ABOUT'))

        expect(global.open).toHaveBeenCalledTimes(1)
        expect(global.open).toHaveBeenCalledWith(GITHUB_URL, '_blank')
    })
})
