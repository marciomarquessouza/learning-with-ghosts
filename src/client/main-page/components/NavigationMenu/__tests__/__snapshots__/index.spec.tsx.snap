// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components >> NavigationMenu should render the NavigationMenu component properly when the user is signed: NavigationMenu-user 1`] = `
<div>
  <div
    class="flex flex-row items-center"
  >
    <div
      class="block md:hidden mx-4"
    >
      <div
        class="relative inline-block text-left"
        data-headlessui-state=""
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class=""
          data-headlessui-state=""
          id="headlessui-menu-button-:r1:"
          type="button"
        >
          <svg
            fill="white"
            height="32px"
            id="Layer_1"
            version="1.1"
            viewBox="0 0 32 32"
          >
            <path
              d="M4,10h24c1.104,0,2-0.896,2-2s-0.896-2-2-2H4C2.896,6,2,6.896,2,8S2.896,10,4,10z M28,14H4c-1.104,0-2,0.896-2,2  s0.896,2,2,2h24c1.104,0,2-0.896,2-2S29.104,14,28,14z M28,22H4c-1.104,0-2,0.896-2,2s0.896,2,2,2h24c1.104,0,2-0.896,2-2  S29.104,22,28,22z"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="hidden md:flex flex-row justify-items-center items-center"
    >
      <div
        class="mx-4"
      >
        <a
          href="#"
        >
          <span
            class="font-josefin font-bold text-lg text-white"
          >
            ABOUT
          </span>
        </a>
      </div>
      <div
        class="mx-4"
      >
        <div
          class="relative inline-block text-left"
          data-headlessui-state=""
        >
          <div>
            <button
              aria-expanded="false"
              aria-haspopup="true"
              class=""
              data-headlessui-state=""
              id="headlessui-menu-button-:r2:"
              type="button"
            >
              <div>
                <span
                  class="inline-flex h-12 w-12 items-center justify-center rounded-full bg-gray-500"
                >
                  <span
                    class="text-lg font-medium leading-none text-white"
                  >
                    U
                  </span>
                </span>
              </div>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`components >> NavigationMenu should render the NavigationMenu component properly: NavigationMenu-default 1`] = `
<div>
  <div
    class="flex flex-row items-center"
  >
    <div
      class="block md:hidden mx-4"
    >
      <div
        class="relative inline-block text-left"
        data-headlessui-state=""
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class=""
          data-headlessui-state=""
          id="headlessui-menu-button-:r0:"
          type="button"
        >
          <svg
            fill="white"
            height="32px"
            id="Layer_1"
            version="1.1"
            viewBox="0 0 32 32"
          >
            <path
              d="M4,10h24c1.104,0,2-0.896,2-2s-0.896-2-2-2H4C2.896,6,2,6.896,2,8S2.896,10,4,10z M28,14H4c-1.104,0-2,0.896-2,2  s0.896,2,2,2h24c1.104,0,2-0.896,2-2S29.104,14,28,14z M28,22H4c-1.104,0-2,0.896-2,2s0.896,2,2,2h24c1.104,0,2-0.896,2-2  S29.104,22,28,22z"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="hidden md:flex flex-row justify-items-center items-center"
    >
      <div
        class="mx-4"
      >
        <a
          href="#"
        >
          <span
            class="font-josefin font-bold text-lg text-white"
          >
            ABOUT
          </span>
        </a>
      </div>
      <div
        class="mx-4"
      >
        <a />
      </div>
    </div>
  </div>
</div>
`;
